rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+
+### Backward Compatibility
+
+The new `db_permissions` field should be automatically ignored by Teleport versions that don't support it.
+
+### Audit Events
+
+Since there are no specific audit events in place for the user creation feature, there is currently no requirement to extend any existing events. However, in the event that audit events are introduced, it would be prudent to contemplate enhancing these events by including a list of the applied permissions. This list would be a combination of pertinent entries in both `spec.allow.db_permissions` and `spec.deny.db_permissions`.
+
+### Observability
+
+The expectation is that permission changes should occur swiftly. If necessary, we may consider monitoring the latency of schema queries and the time required to apply permissions. This becomes particularly relevant when permissions are managed external to the database instance, for instance, through a call to AWS Security Token Service (STS).
+
+To enhance observability, it is advisable to introduce appropriate logging. Debug-level logs can be used to detail individual permissions granted, while keeping in mind that the list might encompass thousands of entries.
+
+### Product Usage
+
+As of the current point in time, there are no plans for the introduction of telemetry.
+
+### Test Plan
+
+In the test plan, a new section should be incorporated, positioned alongside the coverage for the "automated user creation" feature. This section should enumerate and test each supported configuration separately. At the time of writing this RFD, the coverage for the "automated user creation" feature in the test plan is absent, and it should be created.
+
+### Security
+
+The introduction of the new feature has no direct impact on the security of Teleport itself. However, it does have implications for the security of connected databases within the supported configurations. For environments requiring heightened security, there may be value in explicitly excluding specific databases from this feature, potentially using a resource label for this purpose.
+
+It's important to recognize that with sufficiently broad permissions, a user might have the potential to elevate their database permissions further, including creating additional users and granting permissions. Given that this feature does not attempt to model the implications of individual database permissions, there is no foolproof mechanism to prevent such excessive permissions. Therefore, it falls to the system administrator to ensure that the granted permissions are kept to a minimum.
```

Comment:
We should make sure that when documenting this feature we highlight that administrators are taking this responsibility.  And that they need understand their databases and the RBAC of their database systems carefully.

Reply:
Agreed. I also wonder about utility vs. security here; we could refuse to implement the ability to grant certain permissions, to avoid potential footguns, but that might mean users who need those are left out. We might also gate this behind a resource label or other mechanism (to avoid stumbling into this without knowing), but that would necessarily complicate the UX as well as the implementation itself.

Reply:
It may make sense to prevent user creation.  I am not sure if there would be a legitimate need for that considering `teleport` should be doing that management.

I have not thought about this deeply yet, are there other escalation or persistence paths you have considered?

That said, I still think regardless of what limits we put in, documenting this responsibility is important.

Reply:
This feature has a non-trivial consequences for security. Being able to disable it at the resource level (e.g., for all production databases) might be useful: there is no way to misconfigure a disabled feature. I suppose it is somewhat similar in purpose to permission boundaries https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html.